### ISSUE TITLE
Some tests have bugs

### ISSUE THREAD

**Issue by dyfer** [supercollider/supercollider#5359]
## Environment

* SuperCollider version: current develop
* Operating system: Linux, macOS
* Other details (Qt version, audio driver, etc.): tests running locally or in CI

## Steps to reproduce
- add `supercollider/testsuite/classlibrary` to the interpreter's include paths
- reboot interpreter
- run designated test, e.g. `TestCoreUGens.run`

## Expected vs. actual behavior

After implementing our full test suite in the CI, it turned out that there's a number of tests that are currently broken, either due to bugs in tests, or due to test themselves failing. See #5332 for more information.

When fixing the tests, please remove appropriate entries in [gha_test_run_proto.json](https://github.com/supercollider/supercollider/blob/develop/testsuite/scripts/gha_test_run_proto.json)

Tests with bugs:
- [ ] TestCoreUGens
- [ ] TestDocument
- [ ] TestEnvGen_server.test_zero_gate_n_off_not_sent
- [ ] TestFilterUGens
- [ ] TestFilterUGens
- [ ] TestMixedBundleTester
- [ ] TestNode_Server
- [ ] TestProxyNodeMap
- [ ] TestServer_boot

Tests failing (might be due to error in the test, or the test catching an actual bug):
- [ ] TestNodeProxy.test_copy_nodeMapIsCopied
- [ ] TestNodeProxyBusMapping:test_audiorate_mapping
- [ ] TestNodeProxyBusMapping:test_audiorate_mapping_elasticity
- [ ] TestNodeProxy_Server.test_schedAfterFade_notBeforeQuant
- [ ] TestNodeProxy_Server.test_synthDef_isReleased_afterFree
- [ ] TestPanAz
- [ ] TestSCDoc.test_helpSourceDirs_includedExtensions
- [ ] TestTempoClock.test_nextTimeOnGrid_okAfterMeterChange
- [ ] TestUnitTest.test_setUpClass_already_happened

**Comment by patrickdupuis** [comment]
When working on fixing unittests, is it a good idea to have SC built `Debug`? or, does it not matter?

**Comment by dyfer** [comment]
I don't know...

**Comment by patrickdupuis** [comment]
ok, I'm currently going through the entire collection of tests in the test suite to see whick ones fail on my system. Hopefully, some of them are easy fixes.

**Comment by dyfer** [comment]
All tests in this PR should be relatively easy to fix, unless they actually catch a bug that's more involved. Tests from #5360 will be harder to debug...

**Comment by patrickdupuis** [comment]
#5360 is going to be less fun...

**Comment by patrickdupuis** [comment]
I wrote a little scirpt that generates a list of all UnitTests that can be annotated. Gist: https://gist.github.com/patrickdupuis/960d1fdd90f9651f368ab6c0f5e8a0c4

**Comment by patrickdupuis** [comment]
See #5363 for a small fix to the UnitTest class.

**Comment by patrickdupuis** [comment]
With #5363 applied to `develop`, here are the tests that fail on my system (Fedora 33):
https://gist.github.com/patrickdupuis/e7ca9441f77a0bd97458d8393475581b

**Comment by patrickdupuis** [comment]
@dyfer could you add `TestLinkClock` to the list above? I've noticed that it will fail around 50% of the time over here. Specifically, the problem is `test_settingTempo_maintainsSyncOver20To999BPM`. The error message looks like:

```
a TestLinkClock: test_settingTempo_maintainsSyncOver20To999BPM - link clock should stay in sync with its tempo within the range 20-999 bpm
1 of 2 items in array failed to match. Displaying arrays from index of first failure (1) onwards:
[ 0.754712 ]
! =
[ 0.754595 ]
```

**Comment by dyfer** [comment]
@patrickdupuis done!

**Comment by jamshark70** [comment]
@patrickdupuis Perhaps 1e-04 is demanding too much precision (see `classvar precision = 1e-04;`).

On my Linux machine, I ran it several times and didn't see any failures.

It's entirely likely that sync accuracy may be different in different platforms.

1e-03 (maintaining sync within 1/1000 of a beat) is probably more than enough. 1e-02 might even be acceptable.

**Comment by patrickdupuis** [comment]
@jamshark70  I'll try different precissions and see if the failures stop.

**Comment by dyfer** [comment]
@jamshark70 one comment I have is that we run the testsuite essentially in a virtual machine running with "dummy" audio interface, not optimized for audio, so we might need to give tests more "slack" in general.

**Comment by jamshark70** [comment]
> I'll try different precissions and see if the failures stop.

They should:  0.754712 - 0.754595 = 1.17e-4 so it only just barely missed passing.

Probably precision should be 0.01. There's no need for a unit test failure to mean "well it almost passed but didn't because of environmental factors." I had set it as low as I could on my machine, but I agree with @dyfer that we should now retune for higher tolerance.

**Comment by dyfer** [comment]
Re: LinkClock tests: I tried increasing the precision and while there's no failure due to timing inconsistencies, there are some other mysterious failures, see https://github.com/dyfer/supercollider/runs/2291858618?check_suite_focus=true#step:8:8188
I'm really clueless here. Is this some weird issue with the interpreter itself?

**Comment by jamshark70** [comment]
```
ERROR: Message 'prNumPeersChanged' not understood.
RECEIVER:
Instance of Frame {    (0x55b04b4bd4d8, gc=54, fmt=01, flg=00, set=03)
  indexed slots [7]
      0 : instance of Method CollStream:pos_ (0x55b045979200)
      1 : Integer 0
      2 : Frame (0x55b04b4bd4d8) of CollStream:pos_
      3 : Frame (0x55b04b4bd4d8) of CollStream:pos_
      4 : RawPointer (nil)
      5 : instance of CollStream (0x55b049c85288, size=2, set=2)
      6 : Integer 0
}
```

This sort of thing -- calling a method on an object that has no reason to be referenced in the caller -- is often an interpreter bug: perhaps incorrect handling of stack push/pop, or a GC error. `Frame` or `DebugFrame` almost always indicates this sort of bug, as user-facing classes generally do not use these classes directly. (`DebugFrame` may be legit, if the caller did `getBackTrace` -- but (Test)LinkClock doesn't; `Frame` is strictly internal and I can't think of any reason why LinkClock or its test would be addressing a Frame directly.)

`try` is known to cause stack-unwinding bugs. If we use `try` in the test suite, that could be why you're seeing this.

**Comment by dyfer** [comment]
> try is known to cause stack-unwinding bugs. If we use try in the test suite, that could be why you're seeing this.

Yes, I think we use `try` in testsuite so that an error in one test doesn't stop the whole suite.

Debugging interpreter bugs is way above my head. In the meantime, I was able to reproduce errors and failures in some LinkClock tests: #5421. The errors are a mystery to me as I wasn't looking into LinkClock internals. As for the failures, when they occur they seem to _not_ "barely miss" passing, i.e. it seems to be a different issue than `precision` mentioned [above](https://github.com/supercollider/supercollider/issues/5359#issuecomment-787616723).

**Comment by dyfer** [comment]
I've set this issue to be closed by #7052. Once the tests listed in that PR are fixed, I think this can be closed. The only remaining failures - LinkClock and disabled Duty tests - are filed as separate issues.
